src/server/cluster/incoming_slot_migration.cc

```diff
@@ -53,14 +53,16 @@ class ClusterShardMigration {
         break;
       }
 
-      while (tx_data->opcode == journal::Op::FIN) {
-        VLOG(2) << "Attempt to finalize flow " << source_shard_id_;
+      while (tx_data->opcode == journal::Op::LSN) {
+        VLOG(2) << "Attempt to finalize flow " << source_shard_id_ << " attempt " << tx_data->lsn;
+        last_attempt_.store(tx_data->lsn);
```

Comment:
In cases that FIN opcode was sent and than source node retries to finalize it will continue sending more data and therefore tx_reader.NextTxData in the next line will return nullopt?
if so why didnt we get it when you reproduced the bug? is it due to timeout on socket?

do we still need this logic of calling tx_reader.NextTxData in the next line with your fix?

Reply:
also I am kind of confused with the logic of the blocking counter now as we wait on it in the join but we might decrease it but it could be that we returned in line 63 but its actually not the attempt that we are currently waiting on

Reply:
> if so why didnt we get it when you reproduced the bug? is it due to timeout on socket?

Yes, it will continue sending data. But there is a short period of time after `Dec()` and before `Add()` that it is able to "catch" those earlier `FIN`s.
This is why you must set a pretty aggressive timeout, for this to happen a lot.

> do we still need this logic of calling tx_reader.NextTxData in the next line with your fix?

Yes, we still need it, otherwise we won't be able to continue after a timeout. The overall protocol/design is ok, we just need to confirm that we're waiting on the right attempt on both source and target.



> also I am kind of confused with the logic of the blocking counter now as we wait on it in the join but we might decrease it but it could be that we returned in line 63 but its actually not the attempt that we are currently waiting on

If we return in line 63 it means that the socket was closed, so either the source node decided to finalize, or there's something wrong, but anyway there is no way to continue with the loop because there's nothing to read from